service/storage_service.cc

```diff
@@ -1536,6 +1536,39 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
     co_await _sys_ks.local().set_must_synchronize_topology(false);
 }
 
+// Wait until gossiper observes the given nodes or timeout is reached.
+// Called on initial startup (bootstrap/replace), not during restarts.
+static future<> wait_until_seen_during_boot(
+        const gms::gossiper& g, const std::unordered_set<gms::inet_address>& initial_contact_nodes,
+        db::timeout_clock::time_point timeout, abort_source& as) {
+    logger::rate_limit rate_limit{std::chrono::seconds{5}};
+    while (!as.abort_requested()) {
+        std::vector<gms::inet_address> observed = g.get_endpoints();
+        std::unordered_set<gms::inet_address> observed_set{observed.begin(), observed.end()};
+
+        std::vector<gms::inet_address> diff;
+        diff.reserve(initial_contact_nodes.size());
+        for (auto& n: initial_contact_nodes) {
+            if (!observed_set.contains(n)) {
+                diff.push_back(n);
```

Comment:
Nitpick, you don't really need this list "diff", it doesn't seem to be used. You just need a boolean whether all the contact nodes were observed, or not.

Reply:
I print it for better UX - to make it easier to debug if the wait fails (the user will know which nodes are partitioned and so on).
```cpp
        if (db::timeout_clock::now() >= timeout) {
            auto err = ::format("Timed out waiting for nodes {} to show up in gossip during boot", diff);
            slogger.error("{}", err);
            throw std::runtime_error{std::move(err)};
        }

        slogger.log(log_level::info, rate_limit, "Nodes {} not yet seen in gossip during initial boot", diff);
```

Reply:
Oh, I missed that you printed it in the message. Ok.